FirebaseSegmentation/Sources/SEGNetworkManager.m

```diff
@@ -0,0 +1,223 @@
+
+// HTTP default timeout.
+static NSTimeInterval const kSEGHTTPRequestTimeout = 60;
+
+/// Completion handler invoked by URLSession completion handler.
+typedef void (^URLSessionCompletion)(NSData *data, NSURLResponse *response, NSError *error);
+
+@implementation SEGNetworkManager {
+  FIROptions *_firebaseAppOptions;
+  NSURLSession *_URLSession;
+}
+
+- initWithFIROptions:(FIROptions *)options {
+  self = [super init];
+  if (self) {
+    _firebaseAppOptions = options;
+    _URLSession = [self newURLSession];
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [_URLSession invalidateAndCancel];
+}
+
+- (void)makeAssociationRequestToBackendWithData:
+            (NSDictionary<NSString *, NSString *> *)associationData
+                                          token:(NSString *)token
+                                     completion:(SEGRequestCompletion)completionHandler {
+  // Construct the server URL.
+  NSString *URL = [self constructServerURLWithAssociationData:associationData];
+  if (!URL) {
+    FIRLogError(kFIRLoggerSegmentation, @"I-SEG000020", @"Could not construct backend URL.");
+    completionHandler(NO, @{@"errorDescription" : @"Could not construct backend URL"});
+  }
+
+  FIRLogDebug(kFIRLoggerSegmentation, @"I-SEG000019", @"%@",
+              [NSString stringWithFormat:@"Making config request: %@", URL]);
+
+  // Construct the request data.
+  NSString *customInstallationIdentifier =
+      [associationData objectForKey:kSEGCustomInstallationIdentifierKey];
+  // TODO: Add tests for nil.
+  NSDictionary<NSString *, NSString *> *requestDataDictionary =
+      @{kRequestDataCustomInstallationIdString : customInstallationIdentifier};
+  NSError *error = nil;
+  NSData *requestData = [NSJSONSerialization dataWithJSONObject:requestDataDictionary
+                                                        options:0
+                                                          error:nil];
+  if (!requestData || error) {
+    FIRLogError(kFIRLoggerSegmentation, @"I-SEG000021", @"Could not create request data. %@",
+                error.localizedDescription);
+    completionHandler(NO,
+                      @{@"errorDescription" : @"Could not serialize JSON data for network call."});
+  }
+
+  // Handle NSURLSession completion.
+  __weak SEGNetworkManager *weakSelf = self;
+  [self URLSessionDataTaskWithURL:URL
+                          content:requestData
+                            token:token
+                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+                  SEGNetworkManager *strongSelf = weakSelf;
+                  if (!strongSelf) {
+                    FIRLogError(kFIRLoggerSegmentation, @"I-SEG000022",
+                                @"Internal error making network request.");
+                    completionHandler(
+                        NO, @{@"errorDescription" : @"Internal error making network request."});
```

Comment:
This `errorDescription` key is inconsistent with other completion handler invocations (`errorDescription` vs `ErrorDescription`). Can you use a global constant key instead of a string literal in every completion handler?